File: typescript/src/oss_upload_client.test.ts
import assert from "node:assert/strict"
import { describe, it } from "node:test"

import { buildOssUploadClient } from "./oss_upload_client"

describe("buildOssUploadClient", () => {
  it("uploads base64 content to the external oss endpoint", async () => {
    const client = buildOssUploadClient("https://scanner/", " token ", async (url, method, headers, body) => {
      assert.equal(url, "https://scanner/api/external/oss/uploads")
      assert.equal(method, "POST")
      assert.equal(headers.Authorization, "Bearer token")
      assert.equal(headers["Content-Type"], "application/json")

      const payload = JSON.parse(body)
      assert.deepEqual(payload, {
        bucket: "media-bucket",
        object_key: "podcasts/a.mp3",
        content_base64: "aGVsbG8=",
        content_type: "audio/mpeg",
        metadata: { source: "unit-test" },
      })

      return {
        status: 200,
        text: JSON.stringify({
          code: 0,
          data: {
            bucket: "media-bucket",
            object_key: "podcasts/a.mp3",
            uri: "oss://media-bucket/podcasts/a.mp3",
            access_url: "https://signed.example.com/podcasts/a.mp3",
          },
        }),
      }
    })

    const out = await client.uploadObject({
      bucket: "media-bucket",
      objectKey: "podcasts/a.mp3",
      contentBase64: "aGVsbG8=",
      contentType: "audio/mpeg",
      metadata: { source: "unit-test" },
    })

    assert.equal(out.uri, "oss://media-bucket/podcasts/a.mp3")
  })

  it("encodes bytes for uploadObjectBytes", async () => {
    const client = buildOssUploadClient("https://scanner", "", async (_url, _method, _headers, body) => {
      const payload = JSON.parse(body)
      assert.equal(payload.content_base64, "aGVsbG8=")
      assert.equal(payload.content_type, "")
      assert.deepEqual(payload.metadata, {})
      return { status: 200, text: JSON.stringify({ code: 0, data: { uri: "oss://bucket/key.txt" } }) }
    })

    const out = await client.uploadObjectBytes({
      bucket: "bucket",
      objectKey: "key.txt",
      content: new TextEncoder().encode("hello"),
    })

    assert.equal(out.uri, "oss://bucket/key.txt")
  })

  it("raises on failed upload", async () => {
    const client = buildOssUploadClient("https://scanner", "", async () => ({
      status: 403,
      text: "forbidden",
    }))

    await assert.rejects(
      async () =>
        await client.uploadObject({
          bucket: "bucket",
          objectKey: "key.txt",
          contentBase64: "aGVsbG8=",
        }),
      /oss upload request failed: http 403 forbidden/,
    )
  })
})

File: python/octoarms_sdk/oss_upload_client_test.py
from __future__ import annotations

import json
import unittest

from octoarms_sdk.oss_upload_client import build_oss_upload_client


class BuildOSSUploadClientTests(unittest.TestCase):
    def test_uploads_base64_content_to_external_oss_endpoint(self) -> None:
        def fake_request(url: str, method: str, headers: dict[str, str], body: bytes) -> tuple[int, str]:
            self.assertEqual(url, "https://scanner/api/external/oss/uploads")
            self.assertEqual(method, "POST")
            self.assertEqual(headers.get("Authorization"), "Bearer token")
            self.assertEqual(headers.get("Content-Type"), "application/json")

            payload = json.loads(body.decode("utf-8"))
            self.assertEqual(
                payload,
                {
                    "bucket": "media-bucket",
                    "object_key": "podcasts/a.mp3",
                    "content_base64": "aGVsbG8=",
                    "content_type": "audio/mpeg",
                    "metadata": {"source": "unit-test"},
                },
            )

            return 200, json.dumps(
                {
                    "code": 0,
                    "data": {
                        "bucket": "media-bucket",
                        "object_key": "podcasts/a.mp3",
                        "uri": "oss://media-bucket/podcasts/a.mp3",
                        "access_url": "https://signed.example.com/podcasts/a.mp3",
                    },
                }
            )

        client = build_oss_upload_client("https://scanner/", " token ", request_fn=fake_request)
        out = client.upload_object(
            bucket="media-bucket",
            object_key="podcasts/a.mp3",
            content_base64="aGVsbG8=",
            content_type="audio/mpeg",
            metadata={"source": "unit-test"},
        )

        self.assertEqual(out.get("uri"), "oss://media-bucket/podcasts/a.mp3")

    def test_encodes_bytes_for_upload_object_bytes(self) -> None:
        def fake_request(_url: str, _method: str, _headers: dict[str, str], body: bytes) -> tuple[int, str]:
            payload = json.loads(body.decode("utf-8"))
            self.assertEqual(payload["content_base64"], "aGVsbG8=")
            self.assertEqual(payload["content_type"], "")
            self.assertEqual(payload["metadata"], {})
            return 200, json.dumps({"code": 0, "data": {"uri": "oss://bucket/key.txt"}})

        client = build_oss_upload_client("https://scanner", "", request_fn=fake_request)
        out = client.upload_object_bytes(bucket="bucket", object_key="key.txt", content=b"hello")

        self.assertEqual(out.get("uri"), "oss://bucket/key.txt")

    def test_allows_backend_configured_bucket_by_omitting_bucket(self) -> None:
        def fake_request(_url: str, _method: str, _headers: dict[str, str], body: bytes) -> tuple[int, str]:
            payload = json.loads(body.decode("utf-8"))
            self.assertEqual(payload["bucket"], "")
            self.assertEqual(payload["object_key"], "datasets/demo.json")
            return 200, json.dumps(
                {"code": 0, "data": {"bucket": "dataset-bucket", "uri": "oss://dataset-bucket/datasets/demo.json"}}
            )

        client = build_oss_upload_client("https://scanner", "", request_fn=fake_request)
        out = client.upload_object(
            object_key="datasets/demo.json",
            content_base64="eyJvayI6dHJ1ZX0=",
            content_type="application/json",
        )

        self.assertEqual(out.get("bucket"), "dataset-bucket")

    def test_raises_on_failed_upload(self) -> None:
        def fake_request(_url: str, _method: str, _headers: dict[str, str], _body: bytes) -> tuple[int, str]:
            return 403, "forbidden"

        client = build_oss_upload_client("https://scanner", "", request_fn=fake_request)

        with self.assertRaisesRegex(RuntimeError, "oss upload request failed: http 403 forbidden"):
            client.upload_object(bucket="bucket", object_key="key.txt", content_base64="aGVsbG8=")


if __name__ == "__main__":
    unittest.main()
